core/types/crosslink.go

```diff
@@ -4,51 +4,76 @@ import (
 	"math/big"
 	"sort"
 
-	"github.com/ethereum/go-ethereum/rlp"
-
 	"github.com/ethereum/go-ethereum/common"
-
+	"github.com/ethereum/go-ethereum/rlp"
 	"github.com/harmony-one/harmony/block"
+	"github.com/harmony-one/harmony/shard"
 )
 
 // CrossLink is only used on beacon chain to store the hash links from other shards
+// signature and bitmap correspond to |blockNumber|parentHash| byte array
+// Captial to enable rlp encoding
+// Here we replace header to signatures only, the basic assumption is the committee will not be
+// corrupted during one epoch, which is the same as consensus assumption
 type CrossLink struct {
-	ChainHeader *block.Header
+	ParentHashF  common.Hash
+	BlockNumberF *big.Int
+	SignatureF   [96]byte //aggregated signature
+	BitmapF      []byte   //corresponding bitmap mask for agg signature
+	ShardIDF     uint32   //will be verified with signature on |blockNumber|blockHash| is correct
+	ParentEpochF *big.Int //will be verified with signature on |blockNumber|blockHash| is correct
 }
 
 // NewCrossLink returns a new cross link object
 func NewCrossLink(header *block.Header) CrossLink {
-	return CrossLink{header}
+	if header.Number().Uint64() == 0 {
+		return CrossLink{}
+	}
+	epoch := big.NewInt(0)
+	parentBlockNum := big.NewInt(0)
+	// check if previous header is the last block of an epoch
+	if shard.Schedule.IsLastBlock(header.Number().Uint64() - 1) {
```

Comment:
This doesn't work  after staking. Instead, should really fetch the parent Block and read its Epoch. As there can be gaps between epochs now due to staking epoch synchronization.

Reply:
In this case, I will remove epoch field to make the code cleaner